File: Dot examples/ASP Book Examples/Book Examples/Chapter34/cs/App_Code/Entities/Blog.cs
using System;
using System.Linq;
using System.Data.Linq;
using System.Data.Linq.Mapping;
using System.Collections;
using System.Collections.Generic;
using System.Web;
using System.Web.Configuration;
using System.ServiceModel.Syndication;

/// <summary>
/// Represents a blog entry
/// </summary>
public partial class Blog : EntityBase<Blog>
{
    /// <summary>
    /// Returns a brief blog entry (not including post) with paging 
    /// </summary>
    public static IEnumerable<BriefBlog> SelectBriefByTag(string tagName, int startRowIndex, int maximumRows)
    {
        return SelectBriefByTag(tagName).Skip(startRowIndex).Take(maximumRows); 
    }


    /// <summary>
    /// Get full blog entry (including WebUser and comment count)
    /// </summary>
    public static FullBlog GetFull(int id)
    {
        return Table
        .Where( b => b.Id == id)
        .Join
            (
                WebUser.Table,
                b => b.AuthorUserName,
                w => w.UserName,
                (b, w) => new FullBlog { Id = b.Id, Version = b.Version, IsPinned = b.IsPinned, Title = b.Title, AuthorUserName = b.AuthorUserName, IntroductionText = b.IntroductionText, Post = b.Post, DateCreated = b.DateCreated, FirstName = w.FirstName, LastName = w.LastName, CommentCount = BlogComment.Table.Where(c => c.BlogId == b.Id).Count() }
            )
        .Single();
    }


    /// <summary>
    /// Retrieves brief blog entry filtering by blog tag
    /// </summary>
    public static IEnumerable<BriefBlog> SelectBriefByTag(string tagName)
    {
        return BlogTag.Table
        .Where(t => t.Name == tagName)
        .Select(t => t.BlogId)
        .Join(Table, id => id, b => b.Id, (id, b) => b)
        .Join
            (
                WebUser.Table,
                b => b.AuthorUserName,
                w => w.UserName,
                (b, w) => new BriefBlog { Id = b.Id, Title = b.Title, AuthorUserName = b.AuthorUserName, IntroductionText = b.IntroductionText, DateCreated = b.DateCreated, FirstName = w.FirstName, LastName = w.LastName, CommentCount = BlogComment.Table.Where(c => c.BlogId == b.Id).Count()}
            )
            .OrderByDescending(b => b.Id);
    }


    /// <summary>
    /// Select count of brief blog entries by tag
    /// </summary>
    public static int SelectBriefCountByTag(string tagName)
    {
        return SelectBriefByTag(tagName).Count();
    }



    public static IEnumerable<BriefBlog> SelectBrief(int startRowIndex, int maximumRows)
    {
        return Table
        .Join
            (
                WebUser.Table,
                b => b.AuthorUserName,
                w => w.UserName,
                (b, w) => new BriefBlog {Id = b.Id, IsPinned = b.IsPinned, Title = b.Title, AuthorUserName = b.AuthorUserName, IntroductionText = b.IntroductionText, DateCreated = b.DateCreated, FirstName = w.FirstName, LastName = w.LastName, CommentCount = BlogComment.Table.Where( c => c.BlogId == b.Id).Count() }
            )
            .OrderByDescending(b => b.IsPinned)
            .ThenByDescending(b => b.Id)
            .Skip(startRowIndex).Take(maximumRows);
    }


    /// <summary>
    /// Used for both RSS and Atom feeds
    /// </summary>
    /// <returns></returns>
    public static IEnumerable<SyndicationItem> SelectFeed()
    {
        BlogSection config = (BlogSection)WebConfigurationManager.GetSection("blog");
        string url = config.Url;
        List<Blog> items = Select("Id Desc").ToList();
        return items.Select(b => new SyndicationItem(b.Title, b.IntroductionText, new Uri(String.Format("{0}/Blog/Default.aspx?blogId={1}", url, b.Id)), String.Format("{0}/Blog/Default.aspx?blogId={1}", url, b.Id), b.DateCreated));
    }


    /// <summary>
    /// Where all validation happens
    /// </summary>
    protected override void Validate()
    {
        // Required fields
        if (!ValidationUtility.SatisfiesRequired(Title))
            ValidationErrors.Add("Title", "Required");
        if (!ValidationUtility.SatisfiesRequired(IntroductionText))
            ValidationErrors.Add("IntroductionText", "Required");
        if (!ValidationUtility.SatisfiesRequired(Post))
            ValidationErrors.Add("Post", "Required");                       
    }
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter34/vb/App_Code/Entities/Blog.vb
Imports System
Imports System.Linq
Imports System.Data.Linq
Imports System.Data.Linq.Mapping
Imports System.Collections
Imports System.Collections.Generic
Imports System.Web
Imports System.Web.Configuration
Imports System.ServiceModel.Syndication

''' <summary>
''' Represents a blog entry
''' </summary>
Partial Public Class Blog
     Inherits EntityBase(Of Blog)
    ''' <summary>
    ''' Returns a brief blog entry (not including post) with paging 
    ''' </summary>
    Public Shared Function SelectBriefByTag(ByVal tagName As String, ByVal startRowIndex As Integer, ByVal maximumRows As Integer) As IEnumerable(Of BriefBlog)
        Return SelectBriefByTag(tagName).Skip(startRowIndex).Take(maximumRows)
    End Function


    ''' <summary>
    ''' Get full blog entry (including WebUser and comment count)
    ''' </summary>
    Public Shared Function GetFull(ByVal id As Integer) As FullBlog
        Return Table _
            .Where(Function(b) b.Id = id) _
            .Join(WebUser.Table, Function(b) b.AuthorUserName, Function(w) w.UserName, Function(b, w) New FullBlog() With {.Id = b.Id, .Version = b.Version, .IsPinned = b.IsPinned, .Title = b.Title, .AuthorUserName = b.AuthorUserName, .IntroductionText = b.IntroductionText, .Post = b.Post, .DateCreated = b.DateCreated, .FirstName = w.FirstName, .LastName = w.LastName, .CommentCount = BlogComment.Table.Where(Function(c) c.BlogId = b.Id).Count()}) _
            .Single()
    End Function


    ''' <summary>
    ''' Retrieves brief blog entry filtering by blog tag
    ''' </summary>
    Public Shared Function SelectBriefByTag(ByVal tagName As String) As IEnumerable(Of BriefBlog)
        Return BlogTag.Table _
            .Where(Function(t) t.Name = tagName) _
            .Select(Function(t) t.BlogId) _
            .Join(Table, Function(id) id, Function(b) b.Id, Function(id, b) b) _
            .Join(WebUser.Table, Function(b) b.AuthorUserName, Function(w) w.UserName, Function(b, w) New BriefBlog() With {.Id = b.Id, .Title = b.Title, .AuthorUserName = b.AuthorUserName, .IntroductionText = b.IntroductionText, .DateCreated = b.DateCreated, .FirstName = w.FirstName, .LastName = w.LastName, .CommentCount = BlogComment.Table.Where(Function(c) c.BlogId = b.Id).Count()}) _
            .OrderByDescending(Function(b) b.Id)
    End Function


    ''' <summary>
    ''' Select count of brief blog entries by tag
    ''' </summary>
    Public Shared Function SelectBriefCountByTag(ByVal tagName As String) As Integer
        Return SelectBriefByTag(tagName).Count()
    End Function



    Public Shared Function SelectBrief(ByVal startRowIndex As Integer, ByVal maximumRows As Integer) As IEnumerable(Of BriefBlog)
        Return Table _
            .Join(WebUser.Table, Function(b) b.AuthorUserName, Function(w) w.UserName, Function(b, w) New BriefBlog() With {.Id = b.Id, .IsPinned = b.IsPinned, .Title = b.Title, .AuthorUserName = b.AuthorUserName, .IntroductionText = b.IntroductionText, .DateCreated = b.DateCreated, .FirstName = w.FirstName, .LastName = w.LastName, .CommentCount = BlogComment.Table.Where(Function(c) c.BlogId = b.Id).Count()}) _
            .OrderByDescending(Function(b) b.IsPinned) _
            .ThenByDescending(Function(b) b.Id) _
            .Skip(startRowIndex) _
            .Take(maximumRows)
    End Function


    ''' <summary>
    ''' Used for both RSS and Atom feeds
    ''' </summary>
    ''' <returns></returns>
    Public Shared Function SelectFeed() As IEnumerable(Of SyndicationItem)
        Dim config As BlogSection = CType(WebConfigurationManager.GetSection("blog"), BlogSection)
        Dim url As String = config.Url
        Dim items As List(Of Blog) = [Select]("Id Desc").ToList()
        Return items.Select(Function(b) New SyndicationItem(b.Title, b.IntroductionText, New Uri(String.Format("{0}/Blog/Default.aspx?blogId={1}", url, b.Id)), String.Format("{0}/Blog/Default.aspx?blogId={1}", url, b.Id), b.DateCreated))
    End Function


    ''' <summary>
    ''' Where all validation happens
    ''' </summary>
    Protected Overrides Sub Validate()
        ' Required fields
        If Not ValidationUtility.SatisfiesRequired(Title) Then
            ValidationErrors.Add("Title", "Required")
        End If
        If Not ValidationUtility.SatisfiesRequired(IntroductionText) Then
            ValidationErrors.Add("IntroductionText", "Required")
        End If
        If Not ValidationUtility.SatisfiesRequired(Post) Then
            ValidationErrors.Add("Post", "Required")
        End If
    End Sub
End Class

